### PR TITLE
Add support for waitUntilAvailable

### PR DESCRIPTION
**Pull request by jesup**

### CODE REVIEW

File: neqo-http3/src/client_events.rs [hunk 1]
```diff
@@ -338,6 +342,13 @@ impl Http3ClientEvents {
         }
     }
 
+    pub(crate) fn stream_creatable(&self, stream_type: StreamType) {
+        // No deduplication needed: the transport layer already deduplicates
+        // `SendStreamCreatable` events (via `ConnectionEvents::q.contains()`)
+        // before they reach this point.
```
Comment (mxinden) [inline neqo-http3/src/client_events.rs]: I am not sure I understand this comment. This applies to the `neqo-transport` event queue, not the `neqo-http3` event queue. Needed here?

File: neqo-http3/src/client_events.rs [hunk 2]
```diff
@@ -120,6 +120,10 @@ pub enum Http3ClientEvent {
     PushReset { push_id: PushId, error: AppError },
     /// New stream can be created
     RequestsCreatable,
```
Comment (mxinden) [inline neqo-http3/src/client_events.rs:121]: How about removing `RequestCreatable` in favor of `StreamCreatable`?

File: neqo-http3/src/connection_client.rs [hunk 3]
```diff
@@ -803,7 +803,7 @@ impl Http3Client {
     /// # Errors
     ///
     /// This may return an error if the particular session does not exist
-    /// or the connection is not in the active state.
+    /// or the connection is not in the active state or if we're out of streams.
```
Comment (mxinden) [inline neqo-http3/src/connection_client.rs]:
```suggestion
    /// or the connection is not in the active state or it would exceed the connection stream limit.
```

Preference for a neutral phrasing.

File: neqo-http3/src/client_events.rs (hunk 2 above)
Comment (jesup) [reply neqo-http3/src/client_events.rs:121]: Good idea

**Comment by jesup** [comment]
This PR is part of a stack of 13 bookmarks:

1. `trunk()`
1. [users/jesup/stats](https://github.com/mozilla/neqo/pull/3468)
1. [users/jesup/cc_stats](https://github.com/mozilla/neqo/pull/3694)
1. [users/jesup/highwatermarks](https://github.com/mozilla/neqo/pull/3469)
1. [users/jesup/sendgroup_sendorder_tracking](https://github.com/mozilla/neqo/pull/3470)
1. [users/jesup/stats_overhead](https://github.com/mozilla/neqo/pull/3471)
1. [users/jesup/avoid_PMTUD](https://github.com/mozilla/neqo/pull/3472)
1. [users/jesup/anticipated_streams](https://github.com/mozilla/neqo/pull/3473)
1. **users/jesup/waitUntilAvailable ← this PR**
1. [users/jesup/atomicWrite](https://github.com/mozilla/neqo/pull/3475)
1. [users/jesup/WT_DRAIN_SESSION](https://github.com/mozilla/neqo/pull/3476)
1. [users/jesup/datagram_max_size](https://github.com/mozilla/neqo/pull/3477)
1. [users/jesup/cleanup_asserts](https://github.com/mozilla/neqo/pull/3478)
1. [users/jesup/ietf_draft_15_support](https://github.com/mozilla/neqo/pull/3646)

---
*Created with [jj-stack](https://github.com/keanemind/jj-stack)*

File: neqo-http3/src/client_events.rs [hunk 4]
```diff
@@ -360,7 +357,14 @@ impl Http3ClientEvents {
...
     pub(crate) fn goaway_received(&self) {
-        self.remove(|evt| matches!(evt, Http3ClientEvent::RequestsCreatable));
+        self.remove(|evt| {
+            matches!(
+                evt,
+                Http3ClientEvent::StreamCreatable {
+                    stream_type: StreamType::BiDi
+                }
+            )
+        });
```
Comment (martinthomson) [inline neqo-http3/src/client_events.rs]: The original code here was right, but the new code isn't.  It's subtle, but the distinction is rooted in the way in which the protocol changes with WebTransport.

Before WebTransport, a GOAWAY is a signal that the client can't keep making requests.  The server will answer open requests, but it would prefer not to receive any more.

After WebTransport, a GOAWAY means the same for HTTP requests and WebTransport sessions (which are just HTTP requests).  For WebTransport, GOAWAY means "we'd like you to start closing down sessions, but please finish your business".

We don't know what it means to "complete" a session.  That might entail some amount of stream creation in some interpretations.  Consider a MoQ session where streams correspond to video segments.  In that context, "finishing" might mean watching the rest of a video or completing the video conference.  In that case, a lot of streams might be created.

So, I think that you need to consider that `RequestCreatable` and `StreamCreatable` are distinct concepts.  Maybe we need both.

**Review by martinthomson — DISMISSED** [review]
This seems like an obvious change, but it is not really.  See the big comment.

File: neqo-http3/src/client_events.rs [hunk 5]
```diff
@@ -117,8 +117,12 @@ pub enum Http3ClientEvent {
     /// A push stream was been reset due to a `HttpGeneralProtocol` error.
     /// Most common case are malformed response headers.
     PushReset { push_id: PushId, error: AppError },
-    /// New stream can be created
+    /// New HTTP request or WebTransport session can be created
```
Comment (martinthomson) [inline neqo-http3/src/client_events.rs]:
```suggestion
    /// New HTTP request (including extended connect and WebTransport sessions) can be created
```

File: neqo-http3/tests/webtransport.rs [hunk 6]
```diff
@@ -932,6 +933,105 @@ fn wt_stats_bytes_acknowledged() {
     );
 }
 
+/// Verify that `Http3ClientEvent::StreamCreatable` fires for both bidi and unidi
+/// stream types during connection establishment.  This event is the neqo-http3
+/// signal used by the `waitUntilAvailable` WebTransport feature: when the DOM
+/// side is waiting for stream quota to become available, it listens for this
+/// event to retry queued stream-creation requests.
+#[test]
+fn stream_creatable_event_on_connect() {
```
Comment (martinthomson) [inline neqo-http3/tests/webtransport.rs:924]: Another test for you: this, but with the server configured with zero active streams.

File: neqo-http3/tests/webtransport.rs [hunk 7]
```diff
@@ -932,6 +933,105 @@ fn wt_stats_bytes_acknowledged() {
...
+        "Expected StreamCreatable(UniDi) event during connection"
+    );
+}
+
+/// Verify that `Http3ClientEvent::StreamCreatable` fires with the correct
+/// `stream_type` when stream quota is replenished via a `MAX_STREAMS` frame.
+/// This is the core mechanism behind `waitUntilAvailable`: the client waits for
+/// quota, the server eventually grants more, and neqo-http3 signals readiness.
+#[test]
+fn stream_creatable_event_on_max_streams() {
```
Comment (martinthomson) [inline neqo-http3/tests/webtransport.rs:993]: Add a variant on this test with GOAWAY (so you can test the divergence of the request and stream creatable events).

**Review by martinthomson — CHANGES_REQUESTED** [review]
I'm not seeing where `RequestCreatable` and `StreamCreatable` diverge on their handling after GOAWAY.  I looked, but all I see is the filtering that happens when GOAWAY is received.  If we get more stream credit after GOAWAY, why would that surface `StreamCreatable` and not `RequestCreatable`?

A test would really help my confidence on that point.

**Comment by jesup** [comment]
> I'm not seeing where `RequestCreatable` and `StreamCreatable` diverge on their handling after GOAWAY. I looked, but all I see is the filtering that happens when GOAWAY is received. If we get more stream credit after GOAWAY, why would that surface `StreamCreatable` and not `RequestCreatable`?
> 
> A test would really help my confidence on that point.

Done
